src/query.ts

```diff
@@ -24,13 +34,44 @@ export class Query {
     } else if (typeof hashid === 'object') {
       // It's a complete object to pass on
       if ('params' in hashid) {
-        this.params = hashid['params'] as SearchParameters;
+        this.params = this._hydrate(hashid['params'] as SearchParameters);
       } else {
-        this.params = hashid;
+        this.params = this._hydrate(hashid);
       }
     }
   }
 
+  private _parseSortingOptions(sortParams: RequestSortOptions): RequestSortOptions[] {
+    if (isPlainObject(sortParams) && Object.keys(sortParams).length > 1) {
+      throw new Error('To sort by multiple fields use an Array of Objects');
+    }
+
+    if (!isArray(sortParams)) {
+      if (isPlainObject(sortParams)) {
+        const res: Array<RequestSortOptions> = [];
+        res.push(sortParams);
+        return res;
+      } else if (typeof sortParams === 'string') {
+        const obj: RequestSortOptions = {};
+        obj[sortParams] = Sort.ASC;
+        const res: Array<RequestSortOptions> = [];
+        res.push(obj);
+        return res;
+      } else {
+        return null;
+      }
+    } else {
+      return sortParams as Array<RequestSortOptions>;
+    }
+  }
+
+  private _hydrate(params: SearchParameters): SearchParameters {
```

Comment:
I think this exposes too much of our implementation details. Ideally, you should pass a structure to the Query object and it will decide what to do with it, when creating the object. It also defeats having the query and the params separate as they are, in fact, different "animals". 

Anyway, the cleaning is a great idea and that will be going in next commit.

Reply:
Also, Query is not responsible of building the querystring, but the `encode-params` library and functions that take care of that. Query is agnostic about how we transform the query.

Reply:
- The qs building algorithm was just a side note, I know it's out of scope.
- The query, from the (internal) perspective of the builder is just a parameter, like the hashid. The builder provides an interface to build a query without knowing the structure but you can also hydrate a query if you have the right data structure.

For me, these are just convenient helpers for end users:

```
const base = new Query();
// query.query and query.search are a bit confusing, query.text may be better
base.text('zapatilla');

const q = new Query();
q.hydrate(base.params());
client.search(q);
q.addFilter('brand', 'nike');
client.search(q);
```

Anyway, up to you.

Reply:
I don't really see the advantage, when you can do just that like this:

```
const query = new Query({...));           // Just past the object and get hydrated it instantly

const newQuery = new Query(query);      // Pass the query and get a copy

// And if you want to change all the object 
newQuery.setParameters({...});                // Hydrate as a public function seems redundant to me
```

But I agree that `.search` is confusing, but `.text` seems more confusing as it is too generic. Maybe go the long way and put `.searchTerm` or even `.searchText`?